content/blog/2025-05-13-brilooped/index.md

```diff
@@ -0,0 +1,443 @@
++++
+title = "Brilooped"
+[[extra.authors]]
+name = "Gerardo Garcia Teruel Noriega"
+[[extra.authors]]
+name = "Dev Patel"
++++
+
+# Brilooped
+
+## Abstract
+
+In this project we introduce Briloop, an extension into the [Bril programming language](https://capra.cs.cornell.edu/bril/) that supports structured control flow, this allows bril programs to use if-then-else, while, and block statements instead of relying on jumps and branches. We also implemented Ramsey's [Beyond Relooper](https://dl.acm.org/doi/10.1145/3547621) to translate bril programs into briloop automatically, having a an average performance penalty of 9% when using the core bril benchmarks. Brilooped not only enables programmers to write bril programs with structured control flow, but also enables a translation from Bril into WebAssembly.
+
+## Representing Briloop Programs
+
+Brilooped programs extends the Bril instruction set with the following op codes: while, block, if, break, and continue.
+Brilooped extends the Bril instruction set with new control flow operations that provide structured alternatives to jumps and labels. These new op codes allow programmers to write more readable and maintainable code with familiar control flow constructs like loops and conditionals. This structured control flow can also be useful for specific applications such as WASM.
+
+### `while`
+
+The `while` operation introduces a structured loop into Bril programs:
+
+- It takes exactly one argument: a string containing the name of a boolean variable that serves as the loop condition.
+- It includes an additional `children` field that contains an array of arrays of `Instr` objects representing the loop body.
+  - Note: this is an array of arrays to make the language a little more uniform (children field is used in both while loops and if-then-else statements)
+- The condition is evaluated before each iteration. If true, the body executes; if false, the loop terminates.
+- The instructions within the loop body are responsible for potentially updating the condition variable.
+
+Example usage:
+
+```json
+{
+  "args": [
+    "cond"
+  ],
+  "children": [
+    [
+      {
+        "args": [
+          "a",
+          "b"
+        ],
+        "dest": "temp",
+        "op": "add",
+        "type": "int"
+      },
+      {
+        "args": ["b", "ten"],
+        "dest": "cond",
+        "op": "le",
+        "type": "bool"
+      }
+    ]
+  ],
+  "op": "while"
+},
```

Comment:
JSON listings might be a little easier to read if you "wrap" some lines manually. For example, it can be nice to put entire instructions on one line, like this:

```json
{ "args": ["b", "ten"], "dest": "cond", "op": "le", "type": "bool" }
```

Same with, like, `"args": ["cond"]`. It requires a little manual reformatting vs. what you get out of a raw `json.dump`, but it makes a big difference for the reader.